src/main/java/org/broadinstitute/barclay/argparser/CommandLinePluginDescriptor.java

```diff
@@ -76,103 +110,127 @@
+     * Implementations of this method should minimally validate that all values that have
      * been specified on the command line have a corresponding plugin instance (this will
      * detect a user-specified value for which there is no corresponding plugin class).
      *
-     * @throws CommandLineException if a plugin value has been specified that
-     * has no corresponding plugin instance (i.e., the plugin class corresponding to the name
-     * was not discovered)
+     * @throws CommandLineException if a plugin value has been specified that has no corresponding
+     * plugin instance (i.e., the plugin class corresponding to the name was not discovered)
+     * or the plugins otherwise cannot be resolved per the descriptor's argument resolution policy.
      */
-    public abstract void validateArguments() throws CommandLineException;
+    public abstract void validateAndResolvePlugins() throws CommandLineException;
 
     /**
-     * @return the default plugins used for this instance of this descriptor as a list of Object. Used for
-     * help/doc generation.
+     * @return the default plugins enabled for this descriptor as a list of Object. Used for
+     * help/documentation generation for this descriptor.
+     *
+     * NOTE: calling this method after argument parsing (and thus after {@link #validateAndResolvePlugins}
+     * is called) may return a different list than calling it before parsing, since the resolution
+     * policy is implementation-dependent and may depend on the actual argument specified by the user
+     * on the command line. For example, the user may have specified that a plugin originally included
+     * in the list of defaults be disabled. The corresponding plugin class would be returned by this
+     * mehtod BEFORE parsing, but not after, since it would have been removed by
+     * {@link #validateAndResolvePlugins}.
      */
-    public abstract List<Object> getDefaultInstances();
+    public abstract List<T> getDefaultInstances();
 
     /**
-     * @return an ordered List of actual plugin instances that have been specified on the command
-     * line, in the same order they were obtained/created by {@line #getInstance}).
+     * @return an ordered list containing the plugins that result from resolving all command line
+     * arguments with any default plugins that have been provided to this descriptor. This list
+     * represents the plugins that will actually be used by the consumer. The resolution policy is
+     * descriptor-dependent.
+     *
+     * NOTE: calling this method before argument parsing (and thus before {@link #validateAndResolvePlugins}
+     * has been called) may return a different list than calling it after parsing, since the resolution
+     * policy is implementation-dependent and may depend on the actual arguments specified by the user
+     * on the command line.
      */
-    public abstract List<T> getAllInstances();
+    public abstract List<T> getResolvedInstances();
```

Comment:
GATKReadFilterDescriptor breaks this contract, it returns only those specified on the command line with no resolution between the default filters and user specified filters being performed.

Reply:
Right - this is describing a new contract, and that difference is the essential part of the change. "all" instances is ambiguous. So we're moving to getDefaultInstances and getResolvedInstances. Hopefully less ambiguous.